### ISSUE TITLE
Retry for failed `CompletionStage` executions not working as expected

### ISSUE THREAD

**Issue by thjaeckle** [smallrye/smallrye-fault-tolerance#823]
Hi.

I am currently using smallrye messaging in Kotlin, consuming from Kafka with `CompletionStage` as result of the processor-method.
Starting point is a reactive rest client, also returning a `CompletionStage` with an HTTP call result.
```kotlin
    @Incoming("lorawan")
    @Acknowledgment(Acknowledgment.Strategy.MANUAL)
    @Asynchronous
    @Retry(delay = 20, maxRetries = 5)
    fun processor(kafkaMessage: Message<LoRaWanMessage>): CompletionStage<Void> {
        return someService.resolveType(loraMsg.getId())
            .thenCompose { type ->
               // success-case - proceed with transforming message and sending to some other `emitter`
            }
            .exceptionallyCompose { t ->
              if (t.cause is MappingException) {
                // failure-strategy is "dead-letter-queue" - I want to put messages which cannot be mapped directly into a dead-letter-topic without retrying them ..
                kafkaMessage.nack(t.cause) // returns a CompletionStage as well
              } else {
                // I assume that completing with a failed CompletionStage, the @Retry should kick in
                CompletableFuture.failedStage(t.cause)
              }
            }
    }
```

I assumed that when now e.g. the http client call done in `someService.resolveType(loraMsg.getId())` completes exceptionally - and as a result the returned `CompletionStage` of the `processor` method completes exceptionally - the `@Retry` shall be done.
But that does not happen, instead I get logs:
```
SRMSG00201: Error caught while processing a message in method LoRaUplinkKafkaGateway#processor
```
with stacktrace: 
```
org.jboss.resteasy.reactive.ClientWebApplicationException: Received: 'Not Found, status code 404' when invoking: Rest Client method: 'MyService#resolveType' 	at org.jboss.resteasy.reactive.client.impl.RestClientRequestContext.unwrapException(RestClientRequestContext.java:185)
```

Did I setup something wrongly? Or are my understanding of how the `@Retry` with CompletionStage should work?
Any suggestions?

**Comment by thjaeckle** [comment]
I figured it out - and it is quite obvious if I think about it :D

In addition to `@Retry` you e.g. need `@Fallback` configured to get complete control over what happens after all retries were done unsuccessfully:
```kotlin
    @Incoming("lorawan")
    @Acknowledgment(Acknowledgment.Strategy.MANUAL)
    @Asynchronous
    @Retry(delay = 25, maxRetries = 5)
    @ExponentialBackoff
    @Fallback(fallbackMethod = "fallbackProcessor")
    fun processor(kafkaMessage: Message<LoRaWanMessage>): CompletionStage<Void> {
      ...
    }

    fun fallbackProcessor(kafkaMessage: Message<LoRaWanMessage>, ex: Exception): CompletionStage<Void> {
      ...
      // here, I can e.g. 'nack' the Kafka message:
      return kafkaMessage.nack(ex)
    }
```

Maybe this should be more explicit in the documentation, it was at least not for me.

**Comment by Ladicek** [comment]
You mean something like this? https://smallrye.io/docs/smallrye-fault-tolerance/6.2.2/reference/retry.html#interactions

I'd be happy to make further improvements to the documentation. I admit the documentation is currently fairly extensive and it isn't always obvious where to look for something, but I don't know how to fix that. I'm unfortunately not an expert at writing documentation :-)

**Comment by thjaeckle** [comment]
Yes, it is mentioned down below there.
I assume that I would have expected something to be mentioned at the very top, the "Description" of "@Retry":
> 4. Otherwise, the exception is rethrown.

E.g.:
> 4. Otherwise, the exception is rethrown.
>    If no other strategy is defined (see [Interaction with Other Strategies](https://smallrye.io/docs/smallrye-fault-tolerance/6.2.2/reference/retry.html#interactions)), this could lead to a termination of the process (e.g. if the method was invoked by the application framework and not by user code).

But I see that this is of course depending on the "context" of the method - for my case as an `@Incoming` invoked by SmallRye messaging Kafka, failing the service is the default behavior.

**Comment by Ladicek** [comment]
Yeah, I'd very much like to avoid assuming too much about callers. "The exception is rethrown" should be enough for any competent programmer to understand.

One thing I could do is to rearrange the structure of the reference guides. Currently, they are all structured like this:

- description
- configuration
- metrics
- interactions with other strategies
- extra features

This is good enough when considering each strategy in isolation. However, the interactions with other strategies are important part of the overall fault tolerance behavior, so it probably makes sense to move "interactions with other strategies" right after "description" (or perhaps even _into_ "description"). That could help a little.
